src/jdk.jartool/share/classes/sun/tools/jar/FingerPrint.java

```diff
@@ -232,85 +243,73 @@ public int hashCode() {
+            this.majorVersion = majorVersion; // JDK-8296329: extract major version only
+            this.access = access.flagsMask();
             this.name = name;
             this.nestedClass = name.contains("$");
             this.superName = superName;
             this.publicClass = isPublic(access);
         }
 
         @Override
-        public void visitOuterClass(String owner, String name, String desc) {
-            if (!this.nestedClass) return;
-            this.outerClassName = owner;
-        }
-
-        @Override
-        public void visitInnerClass(String name, String outerName, String innerName,
-                                    int access) {
-            if (!this.nestedClass) return;
-            if (outerName == null) return;
-            if (!this.name.equals(name)) return;
-            if (this.outerClassName == null) this.outerClassName = outerName;
-        }
-
-        @Override
-        public FieldVisitor visitField(int access, String name, String desc,
-                                       String signature, Object value) {
-            if (isPublic(access)) {
-                fields.add(new Field(access, name, desc));
-            }
-            return null;
-        }
-
-        @Override
-        public MethodVisitor visitMethod(int access, String name, String desc,
-                                         String signature, String[] exceptions) {
-            if (isPublic(access)) {
-                Set<String> exceptionSet = new HashSet<>();
-                if (exceptions != null) {
-                    for (String e : exceptions) {
-                        exceptionSet.add(e);
+        public void accept(ClassElement cle) {
+            switch (cle) {
+                case InnerClassesAttribute ica -> {
+                    for (var icm : ica.classes()) {
+                        if (this.nestedClass && icm.outerClass().isPresent()
+                                && this.name.equals(icm.innerClass().asInternalName())
+                                && this.outerClassName == null) {
+                            this.outerClassName = icm.outerClass().get().asInternalName();
+                        }
                     }
                 }
-                // treat type descriptor as a proxy for signature because signature
-                // is usually null, need to strip off the return type though
-                int n;
-                if (desc != null && (n = desc.lastIndexOf(')')) != -1) {
-                    desc = desc.substring(0, n + 1);
-                    methods.add(new Method(access, name, desc, exceptionSet));
+                case FieldModel fm -> {
+                    if (isPublic(fm.flags())) {
+                        fields.add(new Field(fm.flags().flagsMask(),
+                                fm.fieldName().stringValue(),
+                                fm.fieldType().stringValue()));
```

Comment:
I find it more readable if the parameters to the Field constructor is aligned:

```suggestion
                        fields.add(new Field(fm.flags().flagsMask(),
                                             fm.fieldName().stringValue(),
                                             fm.fieldType().stringValue()));
```